src/main/java/net/neoforged/neoforge/resource/ResourcePackLoader.java

```diff
@@ -200,45 +197,81 @@ public static Pack.ResourcesSupplier createPackForMod(IModFileInfo mf) {
+                .map(mf -> "mod/" + mf.getModInfos().get(0).getModId())
                 .toList());
-        ids.add(MOD_DATA_ID);
+        ids.add(packType == PackType.CLIENT_RESOURCES ? MOD_RESOURCES_ID : MOD_DATA_ID);
         return ids;
     }
 
-    public static <V> Comparator<Map.Entry<String, V>> getSorter(PackType packType) {
-        List<String> order = new ArrayList<>();
-        order.add("vanilla");
-        if (packType == PackType.CLIENT_RESOURCES) {
-            order.add(MOD_RESOURCES_ID);
-        } else {
-            order.add(MOD_DATA_ID);
+    /*
+     To work with the new KnownPack system, a PackRepository should be able to construct a set of specific packs
+     given a list of names, and be guaranteed that the final list contains all of those name. At the same time, any
+     child packs detected at the root level, while their parent pack is also being added, ought to be included only
+     as a child of the parent. This method implements this filtering in an efficient manner.
+     */
+    @ApiStatus.Internal
+    public static List<Pack> expandAndRemoveRootChildren(Stream<Pack> packs, Collection<Pack> availablePacks) {
+        List<String> hiddenRootPacks = new ArrayList<>();
+        IntList hiddenRootPackIndexes = new IntArrayList();
+        Set<String> hiddenSubPacks = Sets.newHashSet();
+        for (var pack : availablePacks) {
+            if (pack.isRequired()) {
+                // Remove all children of required packs -- as these will always be present but may not be added till after this method is called
+                pack.getChildren().forEach(p -> hiddenSubPacks.add(p.getId()));
+            }
+        }
+        ArrayList<Pack> outPacks = new ArrayList<>();
+        var iterator = packs.iterator();
+        // We iterate the root packs
+        while (iterator.hasNext()) {
+            var rootPack = iterator.next();
+            if (rootPack.isHidden()) {
+                // If we've already found this as a child, remove it
+                if (hiddenSubPacks.contains(rootPack.getId())) {
+                    continue;
+                }
+                hiddenRootPacks.add(rootPack.getId());
+                hiddenRootPackIndexes.add(outPacks.size());
+            }
+            // Now we actually add the pack and its children, noting what the children are as we go
+            outPacks.add(rootPack);
+            for (var pack : rootPack.getChildren()) {
+                if (pack.isHidden()) {
+                    hiddenSubPacks.add(pack.getId());
+                }
+                outPacks.add(pack);
+            }
+        }
+
+        // Then we remove by index, from end to start, any root-level hidden packs that were later found to actually be children
+        for (int i = hiddenRootPackIndexes.size() - 1; i >= 0; i--) {
+            if (hiddenSubPacks.contains(hiddenRootPacks.get(i))) {
+                int index = hiddenRootPackIndexes.getInt(i);
```

Comment:
Needing to track indices like this feels wrong for high-level code. Might it be better to use a linked object hash set and then convert to a list at the end? Might even be faster since you won't be removing from a list by index anymore.

Reply:
The issue is that we don't want an object to be at the last index it occurs at, we want it to be at the index it occurs while nested, unless it never occurs while nested

Reply:
Hmm, actually I can probably get that to work with two sets, can't I.  Lemme poke it and see what I can come up with

Reply:
Got this simplified, thanks to the new sequenced collection stuff we've got in java 21. Lemme know what you think.

Reply:
The new code seems much more reasonable, thanks.